### PR TITLE
[MM35712] MM Apps - Invalidate bindings on oAuth disconnect

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This ticket triggers a refresh bindings websocket event whenever `StoreOAuth2User` app service is called.

`StoreOAuth2User` is accessed through the rest API endpoint `/oauth2/user/<appid>` and is called by an app after a connect completion or disconnect.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-35712

### CODE REVIEW

**Comment by jfrerich** [comment]
@DHaussermann I tested this out using zendesk and the description provided in the Jira ticket.

* /zendesk connect
* /zendesk disconnect
* /zendesk subscribe  (this will no longer show up with this code change)

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- On oath disconnect event all binding  only available to connect user are removed without the need to change channels 
- Tested with binding in command, Channel header and Post menu
- Tested for all applicable bindings in Service Now and Zendesk
- No issues found 
LGTM!

Will retest again if there any changes requested during Dev review. 
Thanks @jfrerich This PR makes for a very nice usability improvement 🎉

**Comment by mickmister** [comment]
Does this result in two binding refresh websocket events being sent to the client when the user connects?

**Comment by jfrerich** [comment]
@mickmister, yes, it does because of the `p.dispatchRefreshBindingsEvent(actingUserID)` at the end of the `CompleteRemoteOAuth2` function.  That makes sense because we previously refreshed the bindings for the connect completion.  I would suggest refreshing the bindings at the AppService level and dropping the direct coupling to the restapi. The refresh could also be pushed into the `SaveUser` function inside `store/oauth2.go` 

https://github.com/mattermost/mattermost-plugin-apps/blob/a6f400df5c56f8fefbe67a14ea41d6b1e2dfffbf/server/proxy/oauth2.go#L50-L83

@lev, what are your thoughts?

File: server/appservices/oauth2.go [hunk 1]
```diff
@@ -39,6 +42,7 @@ func (a *AppServices) StoreOAuth2User(appID apps.AppID, actingUserID string, ref
 	if err = a.ensureFromUser(actingUserID); err != nil {
 		return err
 	}
+	a.mm.Frontend.PublishWebSocketEvent(config.WebSocketEventRefreshBindings, map[string]interface{}{}, &model.WebsocketBroadcast{UserId: actingUserID})
```
Comment (levb) [inline server/appservices/oauth2.go:45]: we should do it only if the value has changed

**Review by levb — CHANGES_REQUESTED** [review]

File: server/appservices/oauth2.go (hunk 1 above)
Comment (jfrerich) [reply server/appservices/oauth2.go:45]: Thanks for explaining the reasoning in the meeting today.  I will make the adjustment in the PR and add the WIP label for now.  I believe there is a `changed` bit that I can use to determine if a value has changed.

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on @lev's comments

File: server/appservices/oauth2.go (hunk 1 above)
Comment (jfrerich) [reply server/appservices/oauth2.go]: I committed a change that compares the unmarshalled values before publishing the websocket event.  Is there a preference for comparing the bytes instead?

**Comment by levb** [comment]
@jfrerich @hanzei What is the status of this PR?

**Comment by hanzei** [comment]
@levb I'm talking over the PR and working on an update.

**Comment by hanzei** [comment]
CI is broken ATM. https://github.com/mattermost/mattermost-plugin-apps/pull/266 will fix it.

**Comment by hanzei** [comment]
I've updated the PR to only send the websocket event when the OAuth2 user changes. The event is `CompleteRemoteOAuth2` is not longer send as an app will set the OAuth2 user anyway after the authentication flow is complete. I've also updated the example app to make use of the changes and only show the `connect` or `disconnect` command if it's suitable. 

@levb @mickmister Please take another look.

**Comment by jfrerich** [comment]
@hanzei, I pushed a small spelling fix to this PR. Thanks for taking this one.

**Comment by hanzei** [comment]
@mickmister @levb Maybe we can get this into v0.8. Could you please take a look when you have the time?

File: server/proxy/oauth2.go [hunk 2]
```diff
@@ -73,6 +73,5 @@ func (p *Proxy) CompleteRemoteOAuth2(in Incoming, appID apps.AppID, urlValues ma
 
 	p.conf.Telemetry().TrackOAuthComplete(string(appID), in.ActingUserID)
 
-	p.dispatchRefreshBindingsEvent(in.ActingUserID)
```
Comment (levb) [inline server/proxy/oauth2.go:76]: What happened to this function? Is it still used? If it is, is it different from the explicit new `PublishWebSocketEvent` above?

**Review by levb — APPROVED** [review]

File: server/proxy/oauth2.go (hunk 2 above)
Comment (hanzei) [reply server/proxy/oauth2.go:76]: There is not difference between `dispatchRefreshBindingsEvent` and a manual `PublishWebSocketEvent`. `dispatchRefreshBindingsEvent` is still used in the proxy. https://github.com/mattermost/mattermost-plugin-apps/blob/d3ddbe8a43b48fdcfffaa437330b072f71d8631b/server/proxy/bindings.go#L207-L212

**Comment by hanzei** [comment]
Merging with only one review as it's a small change

**Comment by hanzei** [comment]
Documentation update: https://github.com/mattermost/mattermost-developer-documentation/pull/1210

**Pull request merged**